ding/reward_model/count_based_model.py

```diff
@@ -0,0 +1,354 @@
+            self.train_data = []
+            self.opt = optim.RMSprop(
+                self._counter.parameters(),
+                momentum=0.9,
+                weight_decay=0.95,
+                eps=1e-4,
+            )
+
+    def _train(self):
+        train_data: list = self.train_data
+        train_data: torch.Tensor = torch.stack(train_data).to(self.device)
+        flattened_logits, target_pixel_loss, _ = self._counter(train_data)
+        loss = nn.CrossEntropyLoss()(flattened_logits, target_pixel_loss.long())
+        loss = loss.mean()
+        self.opt.zero_grad()
+        loss.backward()
+        self.opt.step()
+
+    def train(self) -> None:
+        self._train()
+        # self.clear_data()
+
+    def estimate(self, data, t) -> None:
+        """
+            # TODO:
+        """
+        if self._counter_type == 'PixelCNN':
+            obs = self._collect_states(data)
+            obs = torch.stack(obs).to(self.device)
+
+            prob = (self._probs(obs) + 1e-8).sum().item()
+
+            with torch.no_grad():
+                recoding_prob = (self._probs(obs) + 1e-8).sum().item()
+
+            pred_gain = np.log(recoding_prob) - np.log(prob)
+
+            intrinsic_reward = [pow(
+                exp(self.cfg.bonus_coeffient * pow(t+1, -0.5) * pred_gain) - 1, 0.5
+            )]
+
+            for item, rew in zip(data, intrinsic_reward):
+                if self.intrinsic_reward_type == 'add':
+                    item['reward'] += rew
+                elif self.intrinsic_reward_type == 'new':
+                    item['intrinsic_reward'] = rew
+                elif self.intrinsic_reward_type == 'assign':
+                    item['reward'] = rew
+
+    def _probs(self, obs):
+        _, indexes, target = self._counter(obs)
+        pred_prob = target[self.index_range, indexes.long()]
+        return pred_prob
+
+    def _collect_states(self, data):
+        obs = []
+        for item in data:
+            state = item['obs']
+            state = state.permute(1, 2, 0)
+            state = state.numpy()
+            state = cv2.resize(state, self.obs_shape)
+            state = np.reshape(state, [42, 42, 1])
+            state = torch.Tensor(state).to(self.device)
+            state = state.permute(2, 0, 1)  #CHW
+            # print(f'one data in model:{state.shape}')
+            obs.append(state)
+        return obs
+        
+    def collect_data(self, data: list) -> None:
+        obs = self._collect_states(data)
+        self.train_data.extend(obs)
+
+    def clear_data(self) -> None:
+        self.train_data.clear()
+
+            
+
+
+
+# Borrowed implement
+
+class GatedActivation(nn.Module):
+    """Gated activation function as introduced in [2].
+    The function computes actiation_fn(f) * sigmoid(g). The f and g correspond to the
+    top 1/2 and bottom 1/2 of the input channels.
+    """
+
+    def __init__(self, activation_fn=torch.tanh):
```

Comment:
Use `torch.nn.Tanh()` to replace `torch.tanh`, so we can add `print(model)` to know what is applied in this activation. For the details, please refer to `__repr__` method of `nn.Module`

Reply:
got it, thanks.